components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -2582,6 +2555,117 @@ void JsonRpcService::OnGetSPLTokenAccountBalance(
   std::move(callback).Run(amount, decimals, ui_amount_string,
                           mojom::SolanaProviderError::kSuccess, "");
 }
+
+void JsonRpcService::GetSolTokenMetadata(const std::string& nft_account_address,
+                                         GetSolTokenMetadataCallback callback) {
+  // Derive metadata PDA for the NFT accounts
+  absl::optional<std::string> associated_metadata_account =
+      SolanaKeyring::GetAssociatedMetadataAccount(nft_account_address);
+  if (!associated_metadata_account) {
+    std::move(callback).Run(
+        "", mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+    return;
+  }
+
+  auto account_info_metadata_callback =
+      base::BindOnce(&JsonRpcService::OnGetSolanaAccountInfoTokenMetadata,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  // Call getAccountInfo for the metadata PDA
+  auto account_info_response_callback = base::BindOnce(
+      &JsonRpcService::OnGetSolanaAccountInfo, weak_ptr_factory_.GetWeakPtr(),
+      std::move(account_info_metadata_callback));
+  RequestInternal(solana::getAccountInfo(*associated_metadata_account), true,
+                  network_urls_[mojom::CoinType::SOL],
+                  std::move(account_info_response_callback),
+                  solana::ConverterForGetAccountInfo());
+}
+
+void JsonRpcService::OnGetSolanaAccountInfoTokenMetadata(
+    GetSolTokenMetadataCallback callback,
+    absl::optional<SolanaAccountInfo> account_info,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || !account_info) {
+    std::move(callback).Run("", std::move(error), error_message);
+    return;
+  }
+
+  absl::optional<std::vector<uint8_t>> metadata =
+      base::Base64Decode((*account_info).data);
+  if (!metadata) {
+    std::move(callback).Run("", mojom::SolanaProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  absl::optional<GURL> url = DecodeMetadataUri(*metadata);
```

Comment:
From the looks of it this call end up leading to us parsing untrusted inputs as binary arrays, inside this function so we should be doing this on a separate thread or with a memory safe language.